File: apps/desktop-ui/src/lib/preview/calendar.ts
/**
 * داده‌ی نمونه‌ی تقویم سه‌گانه برای پیش‌نمایش مرورگر.
 *
 * جدول مناسبت‌ها آینه‌ی **یک‌به‌یک** همان جدول `novin_core::occasions` است؛
 * تست `audit12_calendar` برابری دو جدول را نگه می‌دارند تا واگرا نشوند.
 * محاسبه‌ی قمری از `lib/hijri` (آینه‌ی الگوریتم هسته) انجام می‌شود.
 */

import {toHijri} from '../hijri'
import {toJalali} from '../format'
import {daysInJalaliMonth, jalaliToDate} from '../dateRange'

export type PreviewOccasion = {
  calendar: 'jalali' | 'hijri'
  month: number
  day: number
  title: string
  holiday: boolean
}

/** مناسبت‌ها — همگام با crates/novin-core/src/occasions.rs */
export const PREVIEW_OCCASIONS: PreviewOccasion[] = [
  {calendar: 'jalali', month: 1, day: 1, title: 'نوروز', holiday: true},
  {calendar: 'jalali', month: 1, day: 2, title: 'عید نوروز', holiday: true},
  {calendar: 'jalali', month: 1, day: 3, title: 'عید نوروز', holiday: true},
  {calendar: 'jalali', month: 1, day: 4, title: 'عید نوروز', holiday: true},
  {calendar: 'jalali', month: 1, day: 12, title: 'روز جمهوری اسلامی', holiday: true},
  {calendar: 'jalali', month: 1, day: 13, title: 'سیزده‌بدر', holiday: true},
  {calendar: 'jalali', month: 3, day: 14, title: 'رحلت امام خمینی', holiday: true},
  {calendar: 'jalali', month: 3, day: 15, title: 'قیام ۱۵ خرداد', holiday: true},
  {calendar: 'jalali', month: 11, day: 22, title: 'پیروزی انقلاب اسلامی', holiday: true},
  {calendar: 'jalali', month: 12, day: 29, title: 'روز ملی شدن صنعت نفت', holiday: true},
  {calendar: 'hijri', month: 1, day: 9, title: 'تاسوعای حسینی', holiday: true},
  {calendar: 'hijri', month: 1, day: 10, title: 'عاشورای حسینی', holiday: true},
  {calendar: 'hijri', month: 2, day: 20, title: 'اربعین حسینی', holiday: true},
  {calendar: 'hijri', month: 2, day: 28, title: 'رحلت رسول اکرم (ص) و شهادت امام حسن مجتبی (ع)', holiday: true},
  {calendar: 'hijri', month: 2, day: 30, title: 'شهادت امام جعفر صادق (ع)', holiday: true},
  {calendar: 'hijri', month: 3, day: 8, title: 'شهادت امام حسن عسکری (ع)', holiday: false},
  {calendar: 'hijri', month: 3, day: 17, title: 'میلاد رسول اکرم (ص) و امام جعفر صادق (ع)', holiday: true},
  {calendar: 'hijri', month: 6, day: 3, title: 'شهادت حضرت فاطمه زهرا (س)', holiday: false},
  {calendar: 'hijri', month: 7, day: 13, title: 'ولادت امام علی (ع)', holiday: false},
  {calendar: 'hijri', month: 7, day: 27, title: 'مبعث رسول اکرم (ص)', holiday: true},
  {calendar: 'hijri', month: 8, day: 15, title: 'ولادت حضرت قائم (عج)', holiday: true},
  {calendar: 'hijri', month: 9, day: 21, title: 'شهادت امام علی (ع)', holiday: true},
  {calendar: 'hijri', month: 10, day: 1, title: 'عید سعید فطر', holiday: true},
  {calendar: 'hijri', month: 10, day: 2, title: 'تعطیل به مناسبت عید سعید فطر', holiday: true},
  {calendar: 'hijri', month: 12, day: 10, title: 'عید سعید قربان', holiday: true},
  {calendar: 'hijri', month: 12, day: 18, title: 'عید سعید غدیر خم', holiday: false},
]

const pad = (value: number) => String(value).padStart(2, '0')
const jalaliString = (date: Date) => {
  const j = toJalali(date)
  return `${j.year}/${pad(j.month)}/${pad(j.day)}`
}
const hijriString = (date: Date) => {
  const h = toHijri(date)
  return `${h.year}/${pad(h.month)}/${pad(h.day)}`
}

function occasionsOf(date: Date) {
  const j = toJalali(date)
  const h = toHijri(date)
  return PREVIEW_OCCASIONS.filter((occasion) =>
    occasion.calendar === 'jalali'
      ? occasion.month === j.month && occasion.day === j.day
      : occasion.month === h.month && occasion.day === h.day,
  ).map((occasion) => ({
    date: date.toISOString().slice(0, 10),
    jalali: jalaliString(date),
    hijri: hijriString(date),
    title: occasion.title,
    calendar: occasion.calendar,
    holiday: occasion.holiday,
  }))
}

/** پاسخ پیش‌نمایش `calendar_overview` — همان قرارداد میزبان. */
export function calendarOverviewResponse(args: Record<string, unknown>) {
  const now = new Date()
  const todayJ = toJalali(now)
  const todayH = toHijri(now)

  let from: Date
  let to: Date
  if (typeof args.fromJalali === 'string' && typeof args.toJalali === 'string') {
    const [fy, fm, fd] = args.fromJalali.split('/').map(Number)
    const [ty, tm, td] = args.toJalali.split('/').map(Number)
    from = jalaliToDate(fy, fm, fd)
    to = jalaliToDate(ty, tm, td)
  } else {
    from = jalaliToDate(todayJ.year, todayJ.month, 1)
    to = jalaliToDate(todayJ.year, todayJ.month, daysInJalaliMonth(todayJ.year, todayJ.month))
  }

  const occasions: ReturnType<typeof occasionsOf> = []
  // پیمایش روزشمار مستقل از منطقه‌ی زمانی — ظهر UTC مبنای امن است.
  const dayCount = Math.round((to.getTime() - from.getTime()) / 86_400_000)
  for (let index = 0; index <= dayCount; index += 1) {
    occasions.push(...occasionsOf(new Date(from.getTime() + index * 86_400_000)))
  }

  const iso = (date: Date) => date.toISOString().slice(0, 10)
  return {
    today: {
      iso: iso(now),
      jalali: jalaliString(now),
      jalali_year: todayJ.year,
      jalali_month: todayJ.month,
      jalali_day: todayJ.day,
      hijri: hijriString(now),
      hijri_year: todayH.year,
      hijri_month: todayH.month,
      hijri_day: todayH.day,
      hijri_month_name: '',
      gregorian: iso(now),
      weekday: (now.getDay() + 1) % 7,
      occasions: occasionsOf(now),
    },
    occasions,
  }
}

File: apps/desktop-host/src-tauri/src/calendar.rs
#![allow(warnings)] // موقت: میراث ممیزی CI — بعد از سبزشدن، فایل‌به‌فایل برداشته می‌شود
//! دستور میزبان تقویم سه‌گانه — شمسی/میلادی/قمری با مناسبت‌ها.
//!
//! همه‌ی محاسبه در `novin_core::{hijri, occasions}` انجام می‌شود تا در CI تست
//! شده باشد؛ این ماژول فقط ورودی شمسی را به تاریخ میلادی می‌برد، خروجی را
//! برای رابط کاربری شکل می‌دهد و بازه را به ۴۰۰ روز محدود می‌کند.
//!
//! تاریخ «امروز» همیشه از ساعت سیستم می‌آید؛ بازه‌ی مناسبت‌ها اختیاری است و
//! پیش‌فرضش ماه شمسیِ جاری است.

use novin_core::jalali::{self, JalaliDate};
use novin_core::occasions::{calendar_day, occasions_between, OccasionDay};
use serde::Serialize;
use tauri::State;

use crate::{require_login, AppState};

#[derive(Debug, Serialize)]
pub struct OccasionJson {
    /// تاریخ میلادی ISO
    pub date: String,
    pub jalali: String,
    pub hijri: String,
    pub title: String,
    /// `jalali` یا `hijri`
    pub calendar: String,
    pub holiday: bool,
}

impl From<OccasionDay> for OccasionJson {
    fn from(day: OccasionDay) -> Self {
        OccasionJson {
            date: novin_core::jalali::iso_string(day.date),
            jalali: day.jalali,
            hijri: day.hijri,
            title: day.title.to_string(),
            calendar: day.calendar.to_string(),
            holiday: day.holiday,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct TodayJson {
    pub iso: String,
    /// `1405/06/02`
    pub jalali: String,
    pub jalali_year: i32,
    pub jalali_month: u32,
    pub jalali_day: u32,
    /// `1448/03/10`
    pub hijri: String,
    pub hijri_year: i32,
    pub hijri_month: u32,
    pub hijri_day: u32,
    pub hijri_month_name: String,
    /// `2026-08-24`
    pub gregorian: String,
    /// شنبه=۰ … جمعه=۶
    pub weekday: u32,
    pub occasions: Vec<OccasionJson>,
}

#[derive(Debug, Serialize)]
pub struct CalendarOverview {
    pub today: TodayJson,
    /// مناسبت‌های بازه به ترتیب تاریخ
    pub occasions: Vec<OccasionJson>,
}

/// نمای تقویم: امروز + مناسبت‌های بازه‌ی دلخواه (پیش‌فرض: ماه شمسی جاری).
#[tauri::command]
pub fn calendar_overview(
    state: State<AppState>,
    from_jalali: Option<String>,
    to_jalali: Option<String>,
) -> Result<CalendarOverview, String> {
    require_login(&state)?;
    let today = chrono::Local::now().date_naive();

    let (from, to) = match (from_jalali, to_jalali) {
        (Some(from), Some(to)) => {
            let from = JalaliDate::parse(&from)
                .and_then(|date| date.to_gregorian())
                .map_err(|error| format!("CAL-001: {error}"))?;
            let to = JalaliDate::parse(&to)
                .and_then(|date| date.to_gregorian())
                .map_err(|error| format!("CAL-001: {error}"))?;
            (from, to)
        }
        _ => {
            // ماه شمسیِ جاری
            let current = jalali::from_gregorian(today);
            let start = JalaliDate::new(current.year, current.month, 1)
                .and_then(|date| date.to_gregorian().ok())
                .ok_or_else(|| "CAL-001: ابتدای ماه شمسی نامعتبر است".to_string())?;
            let end = JalaliDate::new(
                current.year,
                current.month,
                jalali::days_in_jalali_month(current.year, current.month),
            )
            .and_then(|date| date.to_gregorian().ok())
            .ok_or_else(|| "CAL-001: پایان ماه شمسی نامعتبر است".to_string())?;
            (start, end)
        }
    };

    let days = occasions_between(from, to).map_err(|error| format!("CAL-002: {error}"))?;
    let day = calendar_day(today);

    Ok(CalendarOverview {
        today: TodayJson {
            iso: day.gregorian.clone(),
            jalali: day.jalali,
            jalali_year: day.jalali_year,
            jalali_month: day.jalali_month,
            jalali_day: day.jalali_day,
            hijri: day.hijri,
            hijri_year: day.hijri_year,
            hijri_month: day.hijri_month,
            hijri_day: day.hijri_day,
            hijri_month_name: day.hijri_month_name.to_string(),
            gregorian: day.gregorian,
            weekday: day.weekday,
            occasions: day.occasions.into_iter().map(Into::into).collect(),
        },
        occasions: days.into_iter().map(Into::into).collect(),
    })
}
